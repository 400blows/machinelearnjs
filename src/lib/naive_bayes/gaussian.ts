import * as tf from '@tensorflow/tfjs';
import { zip } from 'lodash';
import { reshape, validateFitInputs, validateMatrix2D } from '../ops';
import { IMlModel, Type1DMatrix, Type2DMatrix } from '../types';

const SQRT_2PI = Math.sqrt(Math.PI * 2);

/**
 * The Naive is an intuitive method that uses probabilistic of each attribute
 * being in each class to make a prediction. It uses Gaussian function to estimate
 * probability of a given class.
 *
 * @example
 * import { GaussianNB } from 'kalimdor/naive_bayes';
 *
 * const nb = new GaussianNB();
 * const X = [[1, 20], [2, 21], [3, 22], [4, 22]];
 * const y = [1, 0, 1, 0];
 * nb.fit({ X, y });
 * nb.predict({ X: [[1, 20]] }); // returns [ 1 ]
 *
 */
export class GaussianNB<T extends number | string = number>
  implements IMlModel<T> {
  private classCategories: T[];
  private mean: tf.Tensor2D;
  private variance: tf.Tensor2D;

  /**
   * @param X - array-like or sparse matrix of shape = [n_samples, n_features]
   * @param y - array-like, shape = [n_samples] or [n_samples, n_outputs]
   */
  public fit(X: Type2DMatrix<number> = null, y: Type1DMatrix<T> = null): void {
    validateFitInputs(X, y);
    const { classCategories, mean, variance } = this.fitModel(X, y);
    this.classCategories = classCategories;
    this.mean = mean;
    this.variance = variance;
  }

  /**
   * @param X - array-like, shape = [n_samples, n_features]
   */
  public predict(X: Type2DMatrix<number> = null): T[] {
    validateMatrix2D(X);
    return X.map((x): T => this.singlePredict(x));
  }

  /**
   * Restore the model from saved states
   * @param modelState
   */
  public fromJSON({
    classCategories = null,
    mean = null,
    variance = null
  }: {
    /**
     * List of class categories
     */
    classCategories: T[];
    /**
     * Mean of each feature per class
     */
    mean: Type2DMatrix<number>;
    /**
     * Variance of each feature per class
     */
    variance: Type2DMatrix<number>;
  }): void {
    this.classCategories = classCategories;
    this.mean = tf.tensor2d(mean);
    this.variance = tf.tensor2d(variance);
  }

  /**
   * Save the model's states
   */
  public toJSON(): {
    /**
     * List of class categories
     */
    classCategories: T[];
    /**
     * Mean of each feature per class
     */
    mean: Type2DMatrix<number>;
    /**
     * Variance of each feature per class
     */
    variance: Type2DMatrix<number>;
  } {
    return {
      classCategories: this.classCategories,
      mean: reshape([...this.mean.dataSync()], this.mean.shape) as Type2DMatrix<
        number
      >,
      variance: reshape(
        [...this.variance.dataSync()],
        this.variance.shape
      ) as Type2DMatrix<number>
    };
  }

  /**
   * Make a single prediction
   *
   * @param  {ReadonlyArray<number>} X- values to predict in Matrix format
   * @returns T
   */
  private singlePredict(X: ReadonlyArray<number>): T {
    const matrixX: tf.Tensor<tf.Rank> = tf.tensor1d(X as number[], 'float32');
    const numFeatures = matrixX.shape[0];

    // Comparing input and summary shapes
    const summaryLength = this.mean.shape[1];
    if (numFeatures !== summaryLength) {
      throw new Error(
        `Prediction input ${
          matrixX.shape[0]
        } length must be equal or less than summary length ${summaryLength}`
      );
    }

    const meanValPow: tf.Scalar = matrixX
      .sub(this.mean as tf.Tensor)
      .pow(tf.scalar(2))
      .mul(tf.scalar(-1));

    const exponent: tf.Tensor = meanValPow
      .div(this.variance.mul(tf.scalar(2)))
      .exp() as tf.Tensor;
    const innerDiv: tf.Tensor = tf.scalar(SQRT_2PI).mul(this.variance.sqrt());
    const probabilityArray: tf.Tensor = tf
      .scalar(1)
      .div(innerDiv)
      .mul(exponent);

    const selectionIndex = probabilityArray
      .prod(1)
      .argMax()
      .dataSync()[0];

    return this.classCategories[selectionIndex] as T;
  }

  /**
   * Summarise the dataset per class using "probability density function"
   *
   * @param  {Type2DMatrix<number>} X
   * @param  {ReadonlyArray<T>} y
   * @returns InterfaceFitModel
   */
  private fitModel(
    X: Type2DMatrix<number>,
    y: Type1DMatrix<T>
  ): {
    classCategories: T[];
    mean: tf.Tensor2D;
    variance: tf.Tensor2D;
  } {
    const classCategories = [...new Set(y)].sort() as T[];

    // Separates X by classes specified by y argument
    const separatedByCategory: {
      [key: string]: Type2DMatrix<number>;
    } = zip<ReadonlyArray<number>, T>(X, y).reduce(
      (groups, [row, category]) => {
        groups[category.toString()] = groups[category.toString()] || [];
        groups[category.toString()].push(row);
        return groups;
      },
      {}
    );

    const momentStack = classCategories.map((category: T) => {
      const classFeatures: tf.Tensor = tf.tensor2d(
        separatedByCategory[category.toString()] as number[][],
        null,
        'float32'
      ) as tf.Tensor;
      return tf.moments(classFeatures, [0]);
    });

    // For every class we have a mean and variance for each feature
    const mean: tf.Tensor2D = tf.stack(
      momentStack.map(m => m.mean)
    ) as tf.Tensor2D;
    const variance: tf.Tensor2D = tf.stack(
      momentStack.map(m => m.variance)
    ) as tf.Tensor2D;

    // TODO check for NaN or 0 variance
    // setTimeout(() => {
    //   if ([...variance.dataSync()].some(i => i === 0)) {
    //     console.error('No variance on one of the features. Errors may result.');
    //   }
    // }, 100);

    return {
      classCategories,
      mean,
      variance
    };
  }
}
