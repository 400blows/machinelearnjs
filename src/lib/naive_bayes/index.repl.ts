/* tslint:disable */
import { GaussianNB } from '.';

const nb = new GaussianNB();

const X = [[1, 20], [2, 21], [3, 22], [4, 22]];
const y = [1, 0, 1, 0];
nb.fit(X, y);

const pred = nb.predict([[1, 20]]);
console.log(pred);
