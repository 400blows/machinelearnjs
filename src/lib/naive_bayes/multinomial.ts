import * as tf from '@tensorflow/tfjs';
import { countBy, zip } from 'lodash';
import { reshape, validateFitInputs, validateMatrix2D } from '../ops';
import { IMlModel, Type1DMatrix, Type2DMatrix } from '../types';

/**
 * Multinomial naive bayes machine learning algorithm
 *
 * The Naive is an intuitive method that uses probabilistic of each attribute
 * being in each class to make a prediction. It uses multinomial function to estimate
 * probability of a given class.
 *
 * @example
 * import { MultinomialNB } from 'kalimdor/naive_bayes';
 *
 * const nb = new MultinomialNB();
 * const X = [[1, 20], [2, 21], [3, 22], [4, 22]];
 * const y = [1, 0, 1, 0];
 * nb.fit({ X, y });
 * nb.predict({ X: [[1, 20]] }); // returns [ 1 ]
 *
 */
export class MultinomialNB<T extends number | string = number>
  implements IMlModel<T> {
  /**
   * List of classes
   * @example
   * Given [1, 0, 1, 0, 2, 2, 2], categories are
   * [0, 1, 2]
   */
  private classCategories: T[];
  /**
   * Multinomial distribution values. It is always two dimensional values.
   */
  private multinomialDist: tf.Tensor2D;
  private priorProbability: tf.Tensor1D;

  /**
   *  Setting alpha=1 is called Laplace smoothing, while alpha<1 is called Lidstone smoothing.
   *
   * @param  {number=1} alpha
   */
  constructor(private readonly alpha: number = 1) {}

  /**
   * Fit date to build Gaussian Distribution summary
   *
   * @param  {Type2DMatrix<number>} X - training values
   * @param  {ReadonlyArray<T>} y - target values
   * @returns void
   */
  public fit(X: Type2DMatrix<number>, y: Type1DMatrix<T>): void {
    validateFitInputs(X, y);
    const {
      classCategories,
      multinomialDist,
      priorProbability
    } = this.fitModel(X, y);
    this.classCategories = classCategories as T[];
    this.multinomialDist = multinomialDist;
    this.priorProbability = priorProbability;
  }

  /**
   * Predict multiple rows
   *
   * @param  {Type2DMatrix<number>} X - values to predict in Matrix format
   * @returns T
   */
  public predict(X: Type2DMatrix<number>): T[] {
    validateMatrix2D(X);
    return X.map(x => this.singlePredict(x));
  }

  /**
   * Returns a model checkpoint
   *
   * @returns InterfaceFitModelAsArray
   */
  public toJSON(): {
    classCategories: Type1DMatrix<T>;
    multinomialDist: Type2DMatrix<number>;
    priorProbability: Type1DMatrix<number>;
  } {
    return {
      classCategories: Array.from(this.classCategories),
      priorProbability: Array.from(this.priorProbability.dataSync()),
      multinomialDist: reshape(
        Array.from(this.multinomialDist.dataSync()),
        this.multinomialDist.shape
      ) as Type2DMatrix<number>
    };
  }

  /**
   * @param  {InterfaceFitModelAsArray<T>} modelState
   * @returns void
   */
  public fromJSON(modelState: {
    multinomialDist: Type2DMatrix<number>;
    priorProbability: Type1DMatrix<number>;
    classCategories: Type1DMatrix<T>;
  }): void {
    this.classCategories = modelState.classCategories;
    this.priorProbability = tf.tensor1d(modelState.priorProbability);
    this.multinomialDist = tf.tensor2d(modelState.multinomialDist);
  }

  /**
   * Make a prediction
   *
   * @param  {ReadonlyArray<number>} predictRow
   * @returns T
   */
  private singlePredict(predictRow: Type1DMatrix<number>): T {
    const matrixX = tf.tensor1d(predictRow as number[], 'float32');
    const numFeatures = matrixX.shape[0];
    const summaryLength = this.multinomialDist.shape[1];

    // Comparing input and summary shapes
    if (numFeatures !== summaryLength) {
      throw new Error(
        `Prediction input ${
          matrixX.shape[0]
        } length must be equal or less than summary length ${summaryLength}`
      );
    }

    // log is important to use different multinomial formula instead of the factorial formula
    // The multinomial naive Bayes classifier becomes a linear
    // classifier when expressed in log-space
    // const priorProbability = Math.log(1 / classCount);
    const fitProbabilites = this.multinomialDist
      .clone()
      .mul(matrixX as tf.Tensor);

    // sum(1) is summing columns
    const allProbabilities = fitProbabilites
      .sum(1)
      .add(this.priorProbability as tf.Tensor);

    const selectionIndex = allProbabilities.argMax().dataSync()[0];
    allProbabilities.dispose();

    return this.classCategories[selectionIndex] as T;
  }

  /**
   * Summarise the dataset per class
   *
   * @param  {Type2DMatrix<number>} X - input distribution
   * @param  {ReadonlyArray<T>} y - classes to train
   */
  private fitModel(
    X: Type2DMatrix<number>,
    y: ReadonlyArray<T>
  ): {
    classCategories: T[];
    multinomialDist: tf.Tensor2D;
    priorProbability: tf.Tensor1D;
  } {
    const classCounts = countBy<T>(y);
    const classCategories = Array.from(new Set(y));
    const numFeatures = X[0].length;
    const separatedByCategory = zip<ReadonlyArray<number>, T>(X, y).reduce(
      (groups, [row, category]) => {
        if (!(category.toString() in groups)) {
          groups[category.toString()] = [];
        }
        groups[category.toString()].push(
          tf.tensor1d(row as number[], 'float32')
        );

        return groups;
      },
      {}
    );
    const frequencySumByClass = tf.stack(
      classCategories.map((category: T) =>
        tf.addN(separatedByCategory[category.toString()])
      )
    );
    const productReducedRow = Array.from(frequencySumByClass.sum(1).dataSync());

    // A class's prior may be calculated by assuming equiprobable classes
    // (i.e., priors = (number of samples in the class) / (total number of samples))
    const priorProbability: tf.Tensor1D = tf
      .tensor1d(
        classCategories.map(c => classCounts[c.toString()] / y.length),
        'float32'
      )
      .log();
    // log transform to use linear multinomial forumla
    const multinomialDist: tf.Tensor2D = frequencySumByClass
      .add(tf.scalar(this.alpha) as tf.Tensor)
      .div(
        tf
          .tensor2d(
            productReducedRow as number[],
            [frequencySumByClass.shape[0], 1],
            'float32'
          )
          .add(tf.scalar(numFeatures * this.alpha) as tf.Tensor)
      )
      .log() as tf.Tensor2D;
    return {
      classCategories,
      multinomialDist,
      priorProbability
    };
  }
}
