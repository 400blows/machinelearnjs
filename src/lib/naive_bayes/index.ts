import { installTF } from '../utils/deps';
installTF();
import { GaussianNB } from './gaussian';
import { MultinomialNB } from './multinomial';

export { GaussianNB, MultinomialNB };
