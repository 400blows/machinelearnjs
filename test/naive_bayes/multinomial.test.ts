import { MultinomialNB } from '../../src/lib/naive_bayes/multinomial';

describe('naive_bayes:MultinomialNB', () => {
  const X1 = [[6, 9], [5, 5], [9, 5]];
  const y1 = ['1', '2', '3'];
  const expectedTests: ReadonlyArray<[[number, number], string]> = [
    [[6, 9], '1'],
    [[5, 5], '2'],
    [[9, 5], '3'],
    [[1, 9], '1'],
    [[9, 9], '2'],
    [[90, 9], '3'],
    [[1, 90], '1']
  ];
  const X2 = [[6, 9], [5, 5], [9, 5], [9, 6]];
  const y2 = ['1', '2', '3', '3'];
  const expectedTests2: ReadonlyArray<[[number, number], string]> = [
    [[6, 9], '1'],
    [[5, 5], '3'],
    [[9, 5], '3'],
    [[1, 9], '1'],
    [[9, 9], '3'],
    [[90, 9], '3'],
    [[1, 90], '1']
  ];
  it('Should fit X1 and y1', () => {
    const nb = new MultinomialNB<string>();
    nb.fit(X1, y1);

    expectedTests.forEach(expectedTest => {
      const [test, expected] = expectedTest;
      const result = nb.predict([test]);
      expect(result).toEqual([expected]);
    });

    // test input as tests on fited model
    X1.forEach((test, i) => {
      const result = nb.predict([test]);
      expect(result).toEqual([y1[i]]);
    });
  });
  it('Should fit X1 and y1 as number', () => {
    const nb = new MultinomialNB();
    nb.fit(X1, y1.map(y => +y));

    expectedTests.forEach(expectedTest => {
      const [test, expected] = expectedTest;
      const result = nb.predict([test]);
      expect(result).toEqual([+expected]);
    });
    X1.forEach((test, i) => {
      const result = nb.predict([test]);
      expect(result).toEqual([+y1[i]]);
    });
  });
  it('Should fit X1 and y1 with 2to1 class maps', () => {
    const nb = new MultinomialNB<string>();
    nb.fit(X2, y2);

    expectedTests2.forEach(expectedTest => {
      const [test, expected] = expectedTest;
      const result = nb.predict([test]);
      expect(result).toEqual([expected]);
    });
  });

  it('Should refit X1 and then predict the same', () => {
    const expected = ['1'];

    // Initial model
    const nb = new MultinomialNB<string>();
    nb.fit(X1, y1);
    const result = nb.predict([[1, 20]]);
    expect(result).toEqual(expected);

    nb.fit(X1, y1);
    expect(nb.predict([[1, 20]])).toEqual(expected);
  });
  it('Should fit X1 and y1 and reload then predict the same', () => {
    const expected = ['1'];

    // Initial model
    const nb = new MultinomialNB<string>();
    nb.fit(X1, y1);
    const result = nb.predict([[1, 20]]);
    expect(result).toEqual(expected);

    // Restored model
    const checkpoint = JSON.parse(JSON.stringify(nb.toJSON()));
    const nb2 = new MultinomialNB<string>();
    nb2.fromJSON(checkpoint);
    const result2 = nb2.predict([[1, 20]]);
    expect(result2).toEqual(expected);
  });
});
