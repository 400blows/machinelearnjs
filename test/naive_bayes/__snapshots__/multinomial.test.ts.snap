// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`naive_bayes:MultinomialNB should not fit invalid inputs 1`] = `[Error: validateMatrix2D has received a non-array argument]`;

exports[`naive_bayes:MultinomialNB should not fit invalid inputs 2`] = `[TypeError: Number of labels=1 does not math number of samples=2]`;

exports[`naive_bayes:MultinomialNB should not fit invalid inputs 3`] = `[Error: The matrix is not 2D shaped: [] of [0]]`;

exports[`naive_bayes:MultinomialNB should not predict if the model is not trained yet 1`] = `[TypeError: Cannot read property 'classCategories' of undefined]`;

exports[`naive_bayes:MultinomialNB should not predict invalid inputs 1`] = `[Error: validateMatrix2D has received a non-array argument]`;

exports[`naive_bayes:MultinomialNB should not predict invalid inputs 2`] = `[Error: validateMatrix2D has received a non-array argument]`;

exports[`naive_bayes:MultinomialNB should not predict invalid inputs 3`] = `[Error: validateMatrix2D has received a non-array argument]`;

exports[`naive_bayes:MultinomialNB should not predict invalid inputs 4`] = `[Error: The matrix is not 2D shaped: [1,2] of [2]]`;
